### PR TITLE
add DirSync Control / Search

### PR DESCRIPTION
**Pull request by vetinari**
this implements the dirsync control as described in
https://msdn.microsoft.com/en-us/library/aa366978(v=vs.85).aspx

check ExampleDirSync() in search_test.go how to use

### CODE REVIEW

**Comment by johnweldon** [comment]
@liggitt will you get a chance to review this on the theoretical merits?  I'm pretty out of touch with LDAP stuff for a few years now.  I'll review on technical merits.
@vetinari Thanks !

File: control.go [hunk 1]
```diff
@@ -377,6 +388,26 @@ func DecodeControl(packet *ber.Packet) Control {
 		value.Value = c.Expire
 
 		return c
+	case ControlTypeDirSync:
+		value.Description += " (DirSync)"
+		c := new(ControlDirSync)
+		if value.Value != nil {
+			valueChildren := ber.DecodePacket(value.Data.Bytes())
```
Comment (johnweldon) [inline control.go]: Any reason `DecodePacketErr` is not used and checked here?

File: control.go [hunk 2]
```diff
@@ -377,6 +388,26 @@ func DecodeControl(packet *ber.Packet) Control {
...
+			value.Data.Truncate(0)
+			value.Value = nil
+			value.AppendChild(valueChildren)
+		}
+		value = value.Children[0]
+		value.Description = "DirSync Control Value"
+		value.Children[0].Description = "Flags"
+		value.Children[1].Description = "MaxAttrCnt"
+		value.Children[2].Description = "Cookie"
```
Comment (johnweldon) [inline control.go]: Is this safe; assuming `Children` has at least 3 elements?

File: search_test.go [hunk 3]
```diff
@@ -29,3 +31,48 @@ func TestNewEntry(t *testing.T) {
 		iteration = iteration + 1
 	}
 }
+
+func ExampleDirSync() {
```
Comment (johnweldon) [inline search_test.go]: Are there tests that could be added too?

File: control.go (hunk 1 above)
Comment (vetinari) [inline control.go]: laziness? ;-) this started as a copy of the paging control above...

File: control.go (hunk 2 above)
Comment (vetinari) [inline control.go]: the spec says there must be always 3, the cookie will be an empty string if it's unset (which only happens when the client sends the first time

File: search_test.go (hunk 3 above)
Comment (vetinari) [inline search_test.go]: yes, encoding / decoding in control_test.go

**Comment by vetinari** [comment]
the race fix would also fix the build of PR https://github.com/go-ldap/ldap/pull/104

File: conn.go [hunk 4]
```diff
@@ -96,6 +96,7 @@ type Conn struct {
 	once                sync.Once
 	outstandingRequests uint
 	messageMutex        sync.Mutex
+	timeoutMutex        sync.Mutex
```
Comment (vetinari) [inline conn.go]: see https://travis-ci.org/go-ldap/ldap/jobs/227891683

**Comment by vetinari** [comment]
waiting for #168 ...

**Comment by johnweldon** [comment]
Can we clean this up and get it ready to merge?

**Comment by johnweldon** [comment]
@vetinari any chance you'd want to update this PR too?

**Pull request closed without merging**